server/src/test-fast/java/com/thoughtworks/go/server/service/BuildAssignmentServiceTest.java

```diff
@@ -162,4 +163,69 @@ public void shouldMatchNonElasticJobToNonElasticAgentIfResourcesMatch(){
         verify(elasticAgentPluginService, never()).shouldAssignWork(any(ElasticAgentMetadata.class), any(String.class), any(ElasticProfile.class), any(JobIdentifier.class));
     }
 
+    @Test
+    public void shouldGetMismatchingJobPlansInCaseOfPipelineHasUpdated() {
+        StageConfig second = StageConfigMother.stageConfig("second");
+        second.getJobs().add(JobConfigMother.jobWithNoResourceRequirement());
+
+        PipelineConfig pipeline = PipelineConfigMother.pipelineConfig(UUID.randomUUID().toString());
+        pipeline.get(0).getJobs().add(JobConfigMother.jobWithNoResourceRequirement());
+        pipeline.add(second);
+
+        PipelineConfig irrelevantPipeline = PipelineConfigMother.pipelineConfig(UUID.randomUUID().toString());
+        irrelevantPipeline.get(0).getJobs().add(JobConfigMother.jobWithNoResourceRequirement());
+
+        JobPlan jobPlan1 = getJobPlan(pipeline.getName(), pipeline.get(0).name(), pipeline.get(0).getJobs().last());
+        JobPlan jobPlan2 = getJobPlan(pipeline.getName(), pipeline.get(1).name(), pipeline.get(1).getJobs().first());
+        JobPlan jobPlan3 = getJobPlan(irrelevantPipeline.getName(), irrelevantPipeline.get(0).name(), irrelevantPipeline.get(0).getJobs().first());
+
+        jobPlans.add(jobPlan1);
+        jobPlans.add(jobPlan2);
+        jobPlans.add(jobPlan3);
+
+        //delete a stage
+        pipeline.remove(1);
+
+        List<JobPlan> jobsToRemove = buildAssignmentService.getMismatchingJobPlansFromUpdatedPipeline(pipeline, jobPlans);
```

Comment:
Why dont you invoke `pipelineConfigChangedListener()` from the test instead of exposing these methods - both `getMismatchingJobPlansFromUpdatedPipeline` and `getAllJobPlansFromDeletedPipeline ` for the test. That would make the tests a bit more explicit